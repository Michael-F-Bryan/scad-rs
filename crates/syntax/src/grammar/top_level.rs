use crate::{
    grammar::{expressions, TokenSet},
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const CONTINUE: TokenSet = TokenSet::new([T![;], T!["}"]]);

pub(crate) fn package(p: &mut Parser<'_>) {
    let m = p.start();

    while !p.at(EOF) {
        statement(p);
    }

    p.complete(m, PACKAGE);
}

pub(crate) fn statement(p: &mut Parser<'_>) {
    let m = p.start();

    match p.current() {
        T![include] => include(p, m),
        T![use] => use_(p, m),
        IDENT if p.nth_at(1, T![=]) => {
            assignment(p, m);
            p.expect(T![;]);
        }
        _ => p.error_recover("Expected a statement", m, CONTINUE),
    }
}

pub(crate) fn include(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![include]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, INCLUDE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn use_(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![use]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, USE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn assignment(p: &mut Parser<'_>, m: Mark) {
    p.bump(IDENT);

    p.eat(T![=]);

    expressions::expr(p);
    p.complete(m, ASSIGNMENT);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        empty_package: package(""),
        include_statement: statement(r#"include "./foo/bar""#),
        use_statement: statement(r#"use "./foo/bar""#),
        assignment_statement: statement("x = 42;"),
        kitchen_sink: package(r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#),
    }
}
