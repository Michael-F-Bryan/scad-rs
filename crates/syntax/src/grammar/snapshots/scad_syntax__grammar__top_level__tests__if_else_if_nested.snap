---
source: crates/syntax/src/grammar/top_level.rs
description: "statement(\n            if (test1)\n                if (test2) {\n                    scope2_1();\n                } else {\n                    scope2_2();\n                }\n            else\n                if (test3) {\n                    scope3_1();\n                } else {\n                    scope3_2();\n                })"
---
Node(
    IF_STATEMENT@0..312
      WHITESPACE@0..13 "\n            "
      IF_KW@13..15 "if"
      WHITESPACE@15..16 " "
      L_PAREN@16..17 "("
      LOOKUP_EXPR@17..22
        IDENT@17..22 "test1"
      R_PAREN@22..23 ")"
      IF_STATEMENT@23..159
        WHITESPACE@23..40 "\n                "
        IF_KW@40..42 "if"
        WHITESPACE@42..43 " "
        L_PAREN@43..44 "("
        LOOKUP_EXPR@44..49
          IDENT@44..49 "test2"
        R_PAREN@49..50 ")"
        BRACED_ACTIONS@50..102
          WHITESPACE@50..51 " "
          L_CURLY@51..52 "{"
          MODULE_INSTANTIATION@52..84
            WHITESPACE@52..73 "\n                    "
            IDENT@73..81 "scope2_1"
            L_PAREN@81..82 "("
            ARGUMENTS@82..82
            R_PAREN@82..83 ")"
            SEMICOLON@83..84
              SEMICOLON@83..84 ";"
          WHITESPACE@84..101 "\n                "
          R_CURLY@101..102 "}"
        WHITESPACE@102..103 " "
        ELSE_KW@103..107 "else"
        BRACED_ACTIONS@107..159
          WHITESPACE@107..108 " "
          L_CURLY@108..109 "{"
          MODULE_INSTANTIATION@109..141
            WHITESPACE@109..130 "\n                    "
            IDENT@130..138 "scope2_2"
            L_PAREN@138..139 "("
            ARGUMENTS@139..139
            R_PAREN@139..140 ")"
            SEMICOLON@140..141
              SEMICOLON@140..141 ";"
          WHITESPACE@141..158 "\n                "
          R_CURLY@158..159 "}"
      WHITESPACE@159..172 "\n            "
      ELSE_KW@172..176 "else"
      IF_STATEMENT@176..312
        WHITESPACE@176..193 "\n                "
        IF_KW@193..195 "if"
        WHITESPACE@195..196 " "
        L_PAREN@196..197 "("
        LOOKUP_EXPR@197..202
          IDENT@197..202 "test3"
        R_PAREN@202..203 ")"
        BRACED_ACTIONS@203..255
          WHITESPACE@203..204 " "
          L_CURLY@204..205 "{"
          MODULE_INSTANTIATION@205..237
            WHITESPACE@205..226 "\n                    "
            IDENT@226..234 "scope3_1"
            L_PAREN@234..235 "("
            ARGUMENTS@235..235
            R_PAREN@235..236 ")"
            SEMICOLON@236..237
              SEMICOLON@236..237 ";"
          WHITESPACE@237..254 "\n                "
          R_CURLY@254..255 "}"
        WHITESPACE@255..256 " "
        ELSE_KW@256..260 "else"
        BRACED_ACTIONS@260..312
          WHITESPACE@260..261 " "
          L_CURLY@261..262 "{"
          MODULE_INSTANTIATION@262..294
            WHITESPACE@262..283 "\n                    "
            IDENT@283..291 "scope3_2"
            L_PAREN@291..292 "("
            ARGUMENTS@292..292
            R_PAREN@292..293 ")"
            SEMICOLON@293..294
              SEMICOLON@293..294 ";"
          WHITESPACE@294..311 "\n                "
          R_CURLY@311..312 "}"
    ,
)
